File: all-problems/number-of-islands.cpp

// link : https://www.leetcode.com/problems/number-of-islands

class Solution {
    public:
    bool isIsland(vector<vector<char>>& grid, int i, int j) {

        if (i < 0 || j < 0 || i >= grid.size() || j >= grid[0].size()) return false;

        bool up = true;
        bool down = true;
        bool left = true;
        bool right = true;

        grid[i][j] = '0';

        if (i > 0) {
            if (grid[i-1][j] == '0') up = true;
            else {
                up = isIsland(grid, i-1, j);
            }
        }

        if (j > 0) {
            if (grid[i][j-1] == '0') left = true;
            else {
                left = isIsland(grid, i, j-1);
            }
        }

        if (i < grid.size()-1) {
            if (grid[i+1][j] == '0') down = true;
            else {
                down = isIsland(grid, i+1, j);
            }
        }

        if (j < grid[0].size()-1) {
            if (grid[i][j+1] == '0') right = true;
            else {
                right = isIsland(grid, i, j+1);
            }
        }

        if (up && down && right && left) return true;
        return false;
    }

    int numIslands(vector<vector<char>>& grid) {
        int cnt = 0;
        for (int i=0; i<grid.size(); i++) {
            for (int j=0; j<grid[0].size(); j++) {
                if (grid[i][j] == '1' && isIsland(grid, i, j)) {
                    cnt++;
                }
            }
        }
        return cnt;
    }
};

File: all-problems/number-of-islands.py

# https://leetcode.com/problems/number-of-islands

class Solution:

    def bfs(self, grid: List[List[str]], i: int, j: int):
        m: int = len(grid)
        n: int = len(grid[0])
        queue: List[Tuple[int, int]] = [(i,j)]
        dirs: List[Tuple[int, int]] = [ (0, 1), (1,0), (-1,0), (0, -1) ]

        isIsland: bool = False
        while queue:
            x,y = queue.pop(0)
            grid[x][y] = '0'

            water: int = 0
            for dx,dy in dirs:
                i = x + dx
                j = y + dy
                if i < 0 or j < 0 or i >= m or j >= n or grid[i][j] == '0':
                    water += 1
                elif (i, j) not in queue:
                    queue.append((i, j))
            if water == 4:
                isIsland = True

        return 1 if isIsland else 0


    def numIslands(self, grid: List[List[str]]) -> int:
        m: int = len(grid)
        n: int = len(grid[0])
        cnt: int = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '0':
                    continue
                else:
                    cnt += self.bfs(grid, i, j)
        print()
        return cnt
        
